go/store/datas/database_common.go

```diff
@@ -417,6 +417,78 @@ func (db *database) doCommit(ctx context.Context, datasetID string, commit types
 	return tryCommitErr
 }
 
+func (db *database) Tag(ctx context.Context, ds Dataset, ref types.Ref, opts TagOptions) (Dataset, error) {
+	return db.doHeadUpdate(
+		ctx,
+		ds,
+		func(ds Dataset) error {
+			st, err := NewTag(ctx, ref, opts.Meta)
+
+			if err != nil {
+				return err
+			}
+
+			return db.doTag(ctx, ds.ID(), st)
+		},
+	)
+}
+
+// doTag manages concurrent access the single logical piece of mutable state: the current Root. It uses
+// the same optimistic writing algorithm as doCommit (see above).
+func (db *database) doTag(ctx context.Context, datasetID string, tag types.Struct) error {
+	if is, err := db.validateTag(ctx, tag); err != nil {
+		return err
+	} else if !is {
+		d.Panic("Can't tag using a non-Commit struct (dataset %s)", datasetID)
```

Comment:
Should `validateTag` be `panicIfNotValidTag` or something? It could provide a better error message at least...there are lots of `false` returns there.

`err` is still reasonable to return in the case of i/o errors / non-logic errors.

Reply:
I tried to mirror the existing logic for `Commit()`, but yeah it's not great. will change